rpcutils/blockgate.go

```diff
@@ -0,0 +1,279 @@
+	g.fetchToHeight = height
+}
+
+func (g *BlockGate) NodeHeight() (int64, error) {
+	_, height, err := g.client.GetBestBlock()
+	return height, err
+}
+
+func (g *BlockGate) BestBlockHeight() int64 {
+	g.RLock()
+	defer g.RUnlock()
+	return g.height
+}
+
+func (g *BlockGate) BestBlockHash() (chainhash.Hash, int64, error) {
+	g.RLock()
+	defer g.RUnlock()
+	var err error
+	hash, ok := g.hashAtHeight[g.height]
+	if !ok {
+		err = fmt.Errorf("hash of best block %d not found", g.height)
+	}
+	return hash, g.height, err
+}
+
+func (g *BlockGate) BestBlock() (*dcrutil.Block, error) {
+	g.RLock()
+	defer g.RUnlock()
+	var err error
+	hash, ok := g.hashAtHeight[g.height]
+	if !ok {
+		err = fmt.Errorf("hash of best block %d not found", g.height)
+	}
+	block, ok := g.blockWithHash[hash]
+	if !ok {
+		err = fmt.Errorf("block %d at height %d not found", hash, g.height)
+	}
+	return block, err
+}
+
+func (g *BlockGate) Block(hash chainhash.Hash) (*dcrutil.Block, error) {
+	g.RLock()
+	defer g.RUnlock()
+	var err error
+	block, ok := g.blockWithHash[hash]
+	if !ok {
+		err = fmt.Errorf("block %d not found", hash)
+	}
+	return block, err
+}
+
+func (g *BlockGate) UpdateToBestBlock() (*dcrutil.Block, error) {
+	_, height, err := g.client.GetBestBlock()
+	if err != nil {
+		return nil, fmt.Errorf("GetBestBlockHash failed: %v", err)
+	}
+
+	return g.UpdateToBlock(height)
+}
+
+func (g *BlockGate) UpdateToNextBlock() (*dcrutil.Block, error) {
+	g.Lock()
+	height := g.height + 1
+	g.Unlock()
+	return g.UpdateToBlock(height)
+}
+
+func (g *BlockGate) UpdateToBlock(height int64) (*dcrutil.Block, error) {
+	g.Lock()
+	defer g.Unlock()
+	return g.updateToBlock(height)
+}
+
+func (g *BlockGate) updateToBlock(height int64) (*dcrutil.Block, error) {
+	block, hash, err := GetBlock(height, g.client)
+	if err != nil {
+		return nil, fmt.Errorf("GetBlock (%d) failed: %v", height, err)
+	}
+
+	g.height = height
+	g.hashAtHeight[height] = *hash
+	g.blockWithHash[*hash] = block
+
+	// Push the new block onto the expiration queue, and remove and old ones if
+	// above capacity.
+	g.rotateIn(height, *hash)
+
+	defer func() {
+		go g.signalHeight(height)
+		go g.signalHash(*hash)
+	}()
+
+	return block, nil
+}
+
+func (g *BlockGate) rotateIn(height int64, hash chainhash.Hash) {
+	// Push this new height-hash pair onto the queue
+	g.expireQueue.q = append(g.expireQueue.q, heightHashPair{height, hash})
+	// If above capacity, pop the oldest off
+	if len(g.expireQueue.q) > g.expireQueue.cap {
+		// Pop
+		oldest := g.expireQueue.q[0]
+		g.expireQueue.q = g.expireQueue.q[1:]
+
+		// Remove the dropped height-hash pair from cache maps only if we don't
+		// see it items in the queue with the same height or hash.
+		if !hashInQueue(g.expireQueue, oldest.hash) {
+			delete(g.blockWithHash, oldest.hash)
+		}
+		if !heightInQueue(g.expireQueue, oldest.height) {
+			delete(g.hashAtHeight, oldest.height)
+		}
+	}
+}
+
+func (g *BlockGate) signalHash(blockhash chainhash.Hash) {
+	g.Lock()

```

Comment:
Comment about this lock here and the reason for running this in a deferred goroutine.